solana/restaking/programs/restaking/src/lib.rs

```diff
@@ -68,13 +73,76 @@ pub mod restaking {
             ctx.accounts.token_mint.key()
         );
 
-        staking_params
+        let token_index = staking_params
             .whitelisted_tokens
             .iter()
-            .find(|&&token_mint| token_mint == ctx.accounts.token_mint.key())
+            .enumerate()
+            .find_map(|(index, &token_mint)| {
+                if token_mint == ctx.accounts.token_mint.key() {
+                    return Some(index);
+                }
+                None
+            })
             .ok_or_else(|| error!(ErrorCodes::TokenNotWhitelisted))?;
 
-        staking_params.total_deposited_amount += amount as u128;
+        let token_price_update = &ctx.accounts.token_price_update;
+        let sol_price_update = &ctx.accounts.sol_price_update;
+
+        let sol_price_feed_id = SOL_PRICE_FEED_ID.to_string();
+        let token_feed_id = staking_params
+            .token_oracle_addresses
+            .get(token_index)
+            .unwrap_or(&sol_price_feed_id);
+
+        let final_amount_in_sol = if token_feed_id == &sol_price_feed_id {
+            let (token_price, sol_price) = if cfg!(feature = "mocks") {
+                let feed_id: [u8; 32] = get_feed_id_from_hex(token_feed_id)?;
+                let sol_price = sol_price_update.get_price_unchecked(
+                    &get_feed_id_from_hex(SOL_PRICE_FEED_ID)?,
+                )?;
+                let token_price =
+                    token_price_update.get_price_unchecked(&feed_id)?;
+                (token_price, sol_price)
+            } else {
+                let maximum_age_in_sec: u64 = 30;
+                let feed_id: [u8; 32] = get_feed_id_from_hex(token_feed_id)?;
+                let sol_price = sol_price_update.get_price_no_older_than(
+                    &Clock::get()?,
+                    maximum_age_in_sec,
+                    &get_feed_id_from_hex(SOL_PRICE_FEED_ID)?,
+                )?;
+                let token_price = token_price_update.get_price_no_older_than(
+                    &Clock::get()?,
+                    maximum_age_in_sec,
+                    &feed_id,
+                )?;
+                (token_price, sol_price)
+            };
+
+            let token_decimals = ctx.accounts.token_mint.decimals;
+
+            let amount_in_sol_decimals = (amount *
+                10u64.pow(SOL_DECIMALS as u32)) /
+                10u64.pow(token_decimals as u32);
+
+            let final_amount_in_sol = ((token_price.price *
+                (amount_in_sol_decimals as i64)) /
+                sol_price.price) as u64;
```

Comment:
You can subtract rather than dividing.  More importantly, I’m not
happy with the i64 and u64 conversions.  Those need to be checked
rather than `as`.  We have overflow checks for other operations but
`as` just interprets the bits.

```suggestion
            let decimals = SOL_DECIMALS - ctx.accounts.token_mint.decimals;
            let amount_in_sol_decimals = amount * 10u64.pow(decimals as u32);
            let amount_in_sol_decimals =
                i64::try_from(amount_in_sol_decimals).unwrap();
            let final_amount_in_sol = token_price.price *
                amount_in_sol_decimals /
                sol_price.price;
            let final_amount_in_sol =
                u64::try_from(final_amount_in_sol).unwrap();
```

Reply:
the thing is that the token decimals can be more than solana decimals so i guess we could do smtg like this then
```
let decimals = (SOL_DECIMALS as i64) - (ctx.accounts.token_mint.decimals as i64);
let amount_in_sol_decimals = if decimals < 0 {
     amount / 10u64.pow(abs(decimals))
} else {
     amount * 10u64.pow(decimals)
}
```

Reply:
That makes sense, though prefer `i64::from(...)` to `... as i64`.  I also wonder if we should be concerned with rounding? I guess in this instance truncating is the correct approach.